### ISSUE TITLE
what happens when bench-routes is down?

### ISSUE THREAD

**Issue by aquibbaig** [bench-routes/bench-routes#438]
When the host is down for some reason(mostly if the network is down), bench routes will alert for every route. What we want to have is to add a check beforehand if the problem is with bench routes, then stop the alerting process and send a simple message explaining that the problem lies with bench-routes -> "bench-routes is down".

**Comment by ruddi10** [comment]
@aquibbaig  I would like to take up this issue just want to confirm that by host do you mean the website we are monitoring?

**Comment by aquibbaig** [comment]
@ruddi10 From host I mean the machine on which bench routes is running.

**Comment by devansh42** [comment]
@aquibbaig  I couldn't figure out this issue properly, Could you please add some more explanation with some examples?

**Comment by aquibbaig** [comment]
Let's visualize a situation where you are chilling at your home, monitoring EA Sport's servers for FIFA, and your Wifi goes off. What would bench routes do in this case? With alerting on, it will alert you for each URL you had monitored it for. (generally, there are <10 known EA servers).
What if you are monitoring from a postman collection having thousands of routes? It is expected to alert you for every route "Route is unreachable". But is it the fault of bench-routes that your WiFi is off? In that case, running a network check to not start monitoring will be useful. It can simply alert once that "Couldn't connect to the network".

**Comment by Tushar3099** [comment]
@aquibbaig Thanks for the explaination, now the usecase of this issue is perfectly understood, instead of alerting for each route we should run a check() to diagnose what is the main cause why is it down and give a single alert. But I had a question. How often we are going to run this diagnose?

**Comment by aquibbaig** [comment]
Good question. To be honest, I do not know. The simplest idea that comes to my mind is to use a dedicated go routine that pings a URL such as `google.com` which is not likely to go down often unless you have network problems. When this go routine throws error, you may stop the monitoring.

**Comment by sachdeva-shrey** [comment]
@aquibbaig Can we handle this on the front end itself by using [this](https://www.npmjs.com/package/react-detect-offline) library? We could show an alert snackbar if the connection is down.

**Comment by shivangraina** [comment]
@aquibbaig  checking for internet connection by making a single http request to a url can fail for any number of reasons.Example google.com may be deliberatly blocked by the local network. Is it a good idea to create a health or status endpoint that we can call as a middleware handler?

**Comment by shivangraina** [comment]
@aquibbaig checking for internet connection by making a single http request to a url can fail for any number of reasons.Example google.com may be deliberatly blocked by the local network. Is it a good idea to use a health or status endpoint that we can call(if provided) or a dns query for the service we are requesting (which is not likely to be blocked) , in a middleware handler?

**Comment by aquibbaig** [comment]
> @aquibbaig Can we handle this on the front end itself by using [this](https://www.npmjs.com/package/react-detect-offline) library? We could show an alert snack bar if the connection is down.

It's important to stop monitoring unnecessarily rather than showing alerts to the user, IMO. Rather than running thousands of go-routines that return garbage data(when network is down), its better to run one go-routine that checks network connection to start/stop the monitoring process.

**Comment by aquibbaig** [comment]
I agree! It was just an example I provided, we probably would not "ping google" in real-world scenarios.

> @aquibbaig checking for internet connection by making a single http request to a url can fail for any number of reasons.Example google.com may be deliberatly blocked by the local network. Is it a good idea to use a health or status endpoint that we can call (if provided) or a dns query for the service we are requesting (which is not likely to be blocked) , in a middleware handler?

Do you have such an approach in mind?

**Comment by shivangraina** [comment]
@aquibbaig I was thinking of doing a dns resolution of the the service we want to test on .We can provide the ip and check if resolves to correct hostname or doesn't at all.By this we can ensure the internet connectivity. Also I would like to take up this issue.

**Comment by shivangraina** [comment]
Anyone working on this ?? if not, I would like to work on this issue.

**Comment by shivangraina** [comment]
@aquibbaig I am currently working on this issue, I had a question, Are we supposed to run the dedicated go routine to check the network connection everytime the end point "/query-matrix" is called ?

**Comment by shivangraina** [comment]
@aquibbaig I am currently working on this issue, What currently I am doing is checking network connection everytime the end-point "/query-matrix" is called and then suppose the network connection went down suddenly then the function will call the 'update-monitoring-services-state' and set it to "off" so that it doesn't monitor anymore. Is the above approach correct?
